File: services/firebaseStorage.ts
/**
 * Firebase Cloud Storage service — enhanced edition.
 *
 * Features:
 *   • Resumable uploads with real-time progress callbacks
 *   • Automatic retry with exponential back-off (configurable)
 *   • Client-side file validation (size & MIME type)
 *   • Image compression before upload (optional)
 *   • Download URL caching (LRU, in-memory)
 *   • Batch / multi-file upload helper
 *   • Metadata retrieval
 *   • Date-partitioned folder layout:
 *       receipts/{YYYY-MM-DD}/{registration_id}.png
 *       documents/{YYYY-MM-DD}/{filename}
 *       lab-reports/{YYYY-MM-DD}/{patientId}_{filename}
 */

import {
  ref,
  uploadBytes,
  uploadString,
  uploadBytesResumable,
  getDownloadURL,
  getMetadata,
  deleteObject,
  listAll,
  type StorageReference,
  type UploadResult,
  type UploadTask,
  type UploadMetadata,
  type FullMetadata,
} from 'firebase/storage';
import { storage } from './firebase';

// ═══════════════════════════════════════════════════════════════════════════════
//  TYPES
// ═══════════════════════════════════════════════════════════════════════════════

export interface UploadProgress {
  /** 0 → 1 */
  progress: number;
  /** Bytes transferred so far */
  bytesTransferred: number;
  /** Total bytes */
  totalBytes: number;
  /** Upload state: 'running' | 'paused' | 'success' | 'canceled' | 'error' */
  state: string;
}

export type ProgressCallback = (progress: UploadProgress) => void;

export interface ReceiptUploadResult {
  downloadUrl: string;
  storagePath: string;
}

export interface DocumentUploadResult {
  downloadUrl: string;
  storagePath: string;
  fileName: string;
}

export interface FileMetadata {
  name: string;
  fullPath: string;
  size: number;
  contentType: string;
  timeCreated: string;
  updated: string;
  customMetadata: Record<string, string>;
}

export interface BatchUploadResult {
  succeeded: DocumentUploadResult[];
  failed: { file: File; error: string }[];
}

// ═══════════════════════════════════════════════════════════════════════════════
//  CONFIG / CONSTANTS
// ═══════════════════════════════════════════════════════════════════════════════

/** Maximum upload size in bytes (default 20 MB). */
const MAX_FILE_SIZE = 20 * 1024 * 1024;

/** Allowed MIME types per category. */
const ALLOWED_TYPES: Record<string, string[]> = {
  receipt: ['image/png', 'image/jpeg', 'image/webp'],
  document: [
    'image/png', 'image/jpeg', 'image/webp',
    'application/pdf',
    'application/msword',
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
  ],
  labReport: [
    'image/png', 'image/jpeg', 'image/webp',
    'application/pdf',
  ],
};

const MAX_RETRIES = 3;
const RETRY_BASE_MS = 500;

// ═══════════════════════════════════════════════════════════════════════════════
//  INTERNAL HELPERS
// ═══════════════════════════════════════════════════════════════════════════════

/** YYYY-MM-DD */
function todayPartition(): string {
  return new Date().toISOString().slice(0, 10);
}

function buildRef(folder: string, fileName: string): StorageReference {
  return ref(storage, `${folder}/${todayPartition()}/${fileName}`);
}

function sanitise(name: string): string {
  return name.replace(/[^a-zA-Z0-9._-]/g, '_');
}

/** Sleep helper for retry back-off. */
function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

// ── Validation ───────────────────────────────────────────────────────────────

export class StorageValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'StorageValidationError';
  }
}

function validateFile(file: File | Blob, category: keyof typeof ALLOWED_TYPES): void {
  if (file.size > MAX_FILE_SIZE) {
    const mb = (MAX_FILE_SIZE / (1024 * 1024)).toFixed(0);
    throw new StorageValidationError(`File exceeds maximum size of ${mb} MB.`);
  }
  const allowed = ALLOWED_TYPES[category];
  if (allowed && file.type && !allowed.includes(file.type)) {
    throw new StorageValidationError(
      `File type "${file.type}" is not allowed. Accepted: ${allowed.join(', ')}`,
    );
  }
}

// ── Retry wrapper ────────────────────────────────────────────────────────────

async function withRetry<T>(fn: () => Promise<T>, retries = MAX_RETRIES): Promise<T> {
  let lastError: unknown;
  for (let attempt = 0; attempt <= retries; attempt++) {
    try {
      return await fn();
    } catch (err) {
      lastError = err;
      if (attempt < retries) {
        const delay = RETRY_BASE_MS * 2 ** attempt + Math.random() * 200;
        console.warn(`[FirebaseStorage] Retry ${attempt + 1}/${retries} in ${Math.round(delay)}ms`, err);
        await sleep(delay);
      }
    }
  }
  throw lastError;
}

// ── Image compression ────────────────────────────────────────────────────────

/**
 * Compress an image blob client-side by drawing to a <canvas> and re-encoding.
 * Returns the original blob unchanged for non-image or if compression yields
 * a larger result.
 */
export async function compressImage(
  blob: Blob,
  { maxWidth = 1920, maxHeight = 1920, quality = 0.82 } = {},
): Promise<Blob> {
  if (!blob.type.startsWith('image/')) return blob;

  return new Promise<Blob>((resolve) => {
    const img = new Image();
    const url = URL.createObjectURL(blob);
    img.onload = () => {
      URL.revokeObjectURL(url);
      let { width, height } = img;
      if (width > maxWidth || height > maxHeight) {
        const ratio = Math.min(maxWidth / width, maxHeight / height);
        width = Math.round(width * ratio);
        height = Math.round(height * ratio);
      }
      const canvas = document.createElement('canvas');
      canvas.width = width;
      canvas.height = height;
      const ctx = canvas.getContext('2d')!;
      ctx.drawImage(img, 0, 0, width, height);
      canvas.toBlob(
        (compressed) => {
          // Keep whichever is smaller
          if (compressed && compressed.size < blob.size) {
            resolve(compressed);
          } else {
            resolve(blob);
          }
        },
        'image/webp',
        quality,
      );
    };
    img.onerror = () => {
      URL.revokeObjectURL(url);
      resolve(blob); // fallback to original
    };
    img.src = url;
  });
}

// ── Download URL cache (LRU-ish, max 200 entries) ───────────────────────────

const _urlCache = new Map<string, { url: string; expires: number }>();
const URL_CACHE_TTL_MS = 30 * 60 * 1000; // 30 minutes
const URL_CACHE_MAX = 200;

function getCachedUrl(path: string): string | null {
  const entry = _urlCache.get(path);
  if (!entry) return null;
  if (Date.now() > entry.expires) {
    _urlCache.delete(path);
    return null;
  }
  return entry.url;
}

function setCachedUrl(path: string, url: string): void {
  if (_urlCache.size >= URL_CACHE_MAX) {
    // evict oldest
    const oldest = _urlCache.keys().next().value;
    if (oldest !== undefined) _urlCache.delete(oldest);
  }
  _urlCache.set(path, { url, expires: Date.now() + URL_CACHE_TTL_MS });
}

// ═══════════════════════════════════════════════════════════════════════════════
//  RESUMABLE UPLOAD (with progress)
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Start a resumable upload and return an object that lets callers
 * pause / resume / cancel and observe progress.
 */
export interface ResumableUploadHandle {
  /** The underlying Firebase UploadTask — gives full control. */
  task: UploadTask;
  /** Resolves when the upload finishes. */
  promise: Promise<ReceiptUploadResult>;
  pause: () => boolean;
  resume: () => boolean;
  cancel: () => boolean;
}

export function uploadResumable(
  storageRef: StorageReference,
  data: Blob | Uint8Array | ArrayBuffer,
  metadata: UploadMetadata,
  onProgress?: ProgressCallback,
): ResumableUploadHandle {
  const task = uploadBytesResumable(storageRef, data, metadata);

  const promise = new Promise<ReceiptUploadResult>((resolve, reject) => {
    task.on(
      'state_changed',
      (snap) => {
        onProgress?.({
          progress: snap.totalBytes ? snap.bytesTransferred / snap.totalBytes : 0,
          bytesTransferred: snap.bytesTransferred,
          totalBytes: snap.totalBytes,
          state: snap.state,
        });
      },
      (err) => reject(err),
      async () => {
        const downloadUrl = await getDownloadURL(task.snapshot.ref);
        const storagePath = task.snapshot.ref.fullPath;
        setCachedUrl(storagePath, downloadUrl);
        resolve({ downloadUrl, storagePath });
      },
    );
  });

  return {
    task,
    promise,
    pause: () => task.pause(),
    resume: () => task.resume(),
    cancel: () => task.cancel(),
  };
}

// ═══════════════════════════════════════════════════════════════════════════════
//  RECEIPT UPLOADS
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Upload a receipt blob with progress tracking, validation, optional
 * compression, and automatic retry.
 */
export async function uploadReceiptBlob(
  registrationId: string,
  blob: Blob,
  options?: {
    contentType?: string;
    compress?: boolean;
    onProgress?: ProgressCallback;
  },
): Promise<ReceiptUploadResult> {
  validateFile(blob, 'receipt');
  let data = blob;
  if (options?.compress !== false) {
    data = await compressImage(blob);
  }

  const storageRef = buildRef('receipts', `${registrationId}.png`);
  const meta: UploadMetadata = {
    contentType: options?.contentType ?? data.type ?? 'image/png',
    customMetadata: { registrationId, uploadedAt: new Date().toISOString() },
  };

  if (options?.onProgress) {
    const handle = uploadResumable(storageRef, data, meta, options.onProgress);
    return handle.promise;
  }

  return withRetry(async () => {
    const result: UploadResult = await uploadBytes(storageRef, data, meta);
    const downloadUrl = await getDownloadURL(result.ref);
    setCachedUrl(result.ref.fullPath, downloadUrl);
    return { downloadUrl, storagePath: result.ref.fullPath };
  });
}

/**
 * Upload a receipt from a base-64 data URL with retry.
 */
export async function uploadReceiptDataUrl(
  registrationId: string,
  dataUrl: string,
): Promise<ReceiptUploadResult> {
  const storageRef = buildRef('receipts', `${registrationId}.png`);
  return withRetry(async () => {
    const result = await uploadString(storageRef, dataUrl, 'data_url', {
      customMetadata: { registrationId, uploadedAt: new Date().toISOString() },
    });
    const downloadUrl = await getDownloadURL(result.ref);
    setCachedUrl(result.ref.fullPath, downloadUrl);
    return { downloadUrl, storagePath: result.ref.fullPath };
  });
}

// ═══════════════════════════════════════════════════════════════════════════════
//  DOCUMENT UPLOADS
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Upload a generic document with validation, progress, and retry.
 */
export async function uploadDocument(
  file: File,
  options?: {
    subfolder?: string;
    compress?: boolean;
    onProgress?: ProgressCallback;
  },
): Promise<DocumentUploadResult> {
  validateFile(file, 'document');
  let data: Blob = file;
  if (options?.compress && file.type.startsWith('image/')) {
    data = await compressImage(file);
  }

  const subfolder = options?.subfolder ?? 'documents';
  const safeName = `${Date.now()}_${sanitise(file.name)}`;
  const storageRef = buildRef(subfolder, safeName);
  const meta: UploadMetadata = {
    contentType: file.type,
    customMetadata: { originalName: file.name, uploadedAt: new Date().toISOString() },
  };

  if (options?.onProgress) {
    const handle = uploadResumable(storageRef, data, meta, options.onProgress);
    const res = await handle.promise;
    return { ...res, fileName: file.name };
  }

  return withRetry(async () => {
    const result = await uploadBytes(storageRef, data, meta);
    const downloadUrl = await getDownloadURL(result.ref);
    setCachedUrl(result.ref.fullPath, downloadUrl);
    return { downloadUrl, storagePath: result.ref.fullPath, fileName: file.name };
  });
}

/**
 * Upload a lab-test report with validation, progress, and retry.
 */
export async function uploadLabReport(
  file: File,
  patientId: string,
  options?: {
    compress?: boolean;
    onProgress?: ProgressCallback;
  },
): Promise<DocumentUploadResult> {
  validateFile(file, 'labReport');
  let data: Blob = file;
  if (options?.compress && file.type.startsWith('image/')) {
    data = await compressImage(file);
  }

  const safeName = `${patientId}_${sanitise(file.name)}`;
  const storageRef = buildRef('lab-reports', safeName);
  const meta: UploadMetadata = {
    contentType: file.type,
    customMetadata: { patientId, originalName: file.name, uploadedAt: new Date().toISOString() },
  };

  if (options?.onProgress) {
    const handle = uploadResumable(storageRef, data, meta, options.onProgress);
    const res = await handle.promise;
    return { ...res, fileName: file.name };
  }

  return withRetry(async () => {
    const result = await uploadBytes(storageRef, data, meta);
    const downloadUrl = await getDownloadURL(result.ref);
    setCachedUrl(result.ref.fullPath, downloadUrl);
    return { downloadUrl, storagePath: result.ref.fullPath, fileName: file.name };
  });
}

// ═══════════════════════════════════════════════════════════════════════════════
//  BATCH UPLOAD
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Upload multiple files in parallel with individual progress tracking.
 * Returns a summary of succeeded and failed uploads.
 */
export async function uploadBatch(
  files: File[],
  options?: {
    subfolder?: string;
    compress?: boolean;
    onFileProgress?: (index: number, progress: UploadProgress) => void;
    concurrency?: number;
  },
): Promise<BatchUploadResult> {
  const concurrency = options?.concurrency ?? 3;
  const succeeded: DocumentUploadResult[] = [];
  const failed: { file: File; error: string }[] = [];

  // Process in chunks of `concurrency`
  for (let i = 0; i < files.length; i += concurrency) {
    const chunk = files.slice(i, i + concurrency);
    const results = await Promise.allSettled(
      chunk.map((file, ci) =>
        uploadDocument(file, {
          subfolder: options?.subfolder,
          compress: options?.compress,
          onProgress: options?.onFileProgress
            ? (p) => options.onFileProgress!(i + ci, p)
            : undefined,
        }),
      ),
    );
    results.forEach((r, ci) => {
      if (r.status === 'fulfilled') succeeded.push(r.value);
      else failed.push({ file: chunk[ci], error: String(r.reason) });
    });
  }
  return { succeeded, failed };
}

// ═══════════════════════════════════════════════════════════════════════════════
//  DOWNLOADS & QUERIES
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Get the download URL for a file (uses cache when available).
 */
export async function getFileUrl(storagePath: string): Promise<string> {
  const cached = getCachedUrl(storagePath);
  if (cached) return cached;

  const storageRef = ref(storage, storagePath);
  const url = await getDownloadURL(storageRef);
  setCachedUrl(storagePath, url);
  return url;
}

/**
 * Retrieve full metadata for a stored file.
 */
export async function getFileMetadata(storagePath: string): Promise<FileMetadata> {
  const storageRef = ref(storage, storagePath);
  const meta: FullMetadata = await getMetadata(storageRef);
  return {
    name: meta.name,
    fullPath: meta.fullPath,
    size: meta.size,
    contentType: meta.contentType ?? 'application/octet-stream',
    timeCreated: meta.timeCreated,
    updated: meta.updated,
    customMetadata: (meta.customMetadata as Record<string, string>) ?? {},
  };
}

/**
 * List all receipts for today.
 */
export async function listTodayReceipts(): Promise<string[]> {
  const folderRef = ref(storage, `receipts/${todayPartition()}`);
  const result = await listAll(folderRef);
  return result.items.map((item) => item.fullPath);
}

/**
 * List all files under an arbitrary folder path.
 */
export async function listFiles(folderPath: string): Promise<string[]> {
  const folderRef = ref(storage, folderPath);
  const result = await listAll(folderRef);
  return result.items.map((item) => item.fullPath);
}

/**
 * List files with their download URLs (for display grids, etc.).
 */
export async function listFilesWithUrls(
  folderPath: string,
): Promise<{ path: string; url: string }[]> {
  const paths = await listFiles(folderPath);
  return Promise.all(
    paths.map(async (p) => ({ path: p, url: await getFileUrl(p) })),
  );
}

// ═══════════════════════════════════════════════════════════════════════════════
//  DELETION
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Delete a single file from Cloud Storage.
 */
export async function deleteFile(storagePath: string): Promise<void> {
  const storageRef = ref(storage, storagePath);
  await deleteObject(storageRef);
  _urlCache.delete(storagePath);
}

/**
 * Delete multiple files in parallel.
 */
export async function deleteFiles(paths: string[]): Promise<{ deleted: string[]; errors: string[] }> {
  const deleted: string[] = [];
  const errors: string[] = [];
  await Promise.allSettled(
    paths.map(async (p) => {
      try {
        await deleteFile(p);
        deleted.push(p);
      } catch {
        errors.push(p);
      }
    }),
  );
  return { deleted, errors };
}

File: backend/core/firebase_storage.py
"""
Firebase Admin SDK — enhanced Cloud Storage helpers.

Features beyond basic upload/download:
  • Signed URLs with configurable expiration (instead of make_public)
  • File existence check
  • Copy & move operations
  • Metadata retrieval & update
  • Automatic retry with exponential back-off
  • File-size & content-type validation
  • Batch operations

Place a Firebase service-account JSON file at the path given by the
FIREBASE_SERVICE_ACCOUNT_KEY env-var (defaults to ./firebase-sa-key.json).
Set FIREBASE_STORAGE_BUCKET to your bucket name (e.g. medikisok.firebasestorage.app).
"""

import os
import time
import uuid
import logging
from datetime import datetime, timezone, timedelta
from typing import Optional

import firebase_admin
from firebase_admin import credentials, storage as fb_storage

logger = logging.getLogger("firebase_storage")

# ═══════════════════════════════════════════════════════════════════════════════
#  CONFIG
# ═══════════════════════════════════════════════════════════════════════════════

_SA_KEY_PATH = os.getenv("FIREBASE_SERVICE_ACCOUNT_KEY", "./firebase-sa-key.json")
_STORAGE_BUCKET = os.getenv("FIREBASE_STORAGE_BUCKET", "medikisok.firebasestorage.app")

MAX_FILE_SIZE = 20 * 1024 * 1024  # 20 MB
MAX_RETRIES = 3
RETRY_BASE_SEC = 0.5
SIGNED_URL_EXPIRY_MINS = int(os.getenv("SIGNED_URL_EXPIRY_MINS", "60"))

ALLOWED_CONTENT_TYPES: dict[str, list[str]] = {
    "receipt": ["image/png", "image/jpeg", "image/webp"],
    "document": [
        "image/png", "image/jpeg", "image/webp",
        "application/pdf",
        "application/msword",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    ],
    "lab-report": [
        "image/png", "image/jpeg", "image/webp",
        "application/pdf",
    ],
}

# ═══════════════════════════════════════════════════════════════════════════════
#  INIT
# ═══════════════════════════════════════════════════════════════════════════════

_firebase_app: Optional[firebase_admin.App] = None


def _ensure_initialised() -> firebase_admin.App:
    """Lazy-init so the module can be imported when credentials are absent."""
    global _firebase_app
    if _firebase_app is not None:
        return _firebase_app

    if not os.path.isfile(_SA_KEY_PATH):
        raise RuntimeError(
            f"Firebase service-account key not found at {_SA_KEY_PATH!r}. "
            "Set FIREBASE_SERVICE_ACCOUNT_KEY env var to the correct path."
        )

    cred = credentials.Certificate(_SA_KEY_PATH)
    _firebase_app = firebase_admin.initialize_app(cred, {
        "storageBucket": _STORAGE_BUCKET,
    })
    return _firebase_app


def get_bucket():
    """Return the default Cloud Storage bucket."""
    _ensure_initialised()
    return fb_storage.bucket()


# ═══════════════════════════════════════════════════════════════════════════════
#  INTERNAL HELPERS
# ═══════════════════════════════════════════════════════════════════════════════

def _today() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%d")


def _sanitise(name: str) -> str:
    return name.replace(" ", "_").replace("/", "_")


class StorageValidationError(ValueError):
    """Raised when a file fails size or content-type validation."""
    pass


def validate_upload(
    data: bytes,
    content_type: str,
    category: str = "document",
) -> None:
    """Validate file size and content type before uploading."""
    if len(data) > MAX_FILE_SIZE:
        mb = MAX_FILE_SIZE // (1024 * 1024)
        raise StorageValidationError(f"File exceeds maximum size of {mb} MB.")
    allowed = ALLOWED_CONTENT_TYPES.get(category)
    if allowed and content_type not in allowed:
        raise StorageValidationError(
            f"Content type '{content_type}' not allowed for {category}. "
            f"Accepted: {', '.join(allowed)}"
        )


def _with_retry(fn, retries: int = MAX_RETRIES):
    """Execute `fn()` with exponential back-off retries."""
    last_exc = None
    for attempt in range(retries + 1):
        try:
            return fn()
        except Exception as exc:
            last_exc = exc
            if attempt < retries:
                delay = RETRY_BASE_SEC * (2 ** attempt)
                logger.warning("Retry %d/%d in %.1fs: %s", attempt + 1, retries, delay, exc)
                time.sleep(delay)
    raise last_exc  # type: ignore[misc]


# ═══════════════════════════════════════════════════════════════════════════════
#  CORE UPLOAD
# ═══════════════════════════════════════════════════════════════════════════════

def upload_bytes(
    data: bytes,
    destination_path: str,
    content_type: str = "application/octet-stream",
    metadata: Optional[dict] = None,
    category: str = "document",
) -> str:
    """Upload raw bytes with validation & retry.  Returns a signed download URL."""
    validate_upload(data, content_type, category)

    def _do():
        bucket = get_bucket()
        blob = bucket.blob(destination_path)
        blob.metadata = {**(metadata or {}), "uploadedAt": datetime.now(timezone.utc).isoformat()}
        blob.upload_from_string(data, content_type=content_type)
        # Generate a signed URL instead of making the file world-public
        return blob.generate_signed_url(
            expiration=timedelta(minutes=SIGNED_URL_EXPIRY_MINS),
            method="GET",
        )

    return _with_retry(_do)


# ═══════════════════════════════════════════════════════════════════════════════
#  DOMAIN UPLOADS
# ═══════════════════════════════════════════════════════════════════════════════

def upload_receipt(registration_id: str, data: bytes, content_type: str = "image/png") -> dict:
    """Upload a receipt image and return metadata."""
    path = f"receipts/{_today()}/{registration_id}.png"
    url = upload_bytes(data, path, content_type, {"registrationId": registration_id}, category="receipt")
    return {"download_url": url, "storage_path": path}


def upload_document(
    file_bytes: bytes,
    original_name: str,
    content_type: str,
    subfolder: str = "documents",
) -> dict:
    """Upload a generic document."""
    safe = _sanitise(original_name)
    path = f"{subfolder}/{_today()}/{uuid.uuid4().hex[:8]}_{safe}"
    url = upload_bytes(file_bytes, path, content_type, {"originalName": original_name}, category="document")
    return {"download_url": url, "storage_path": path, "file_name": original_name}


def upload_lab_report(
    file_bytes: bytes,
    original_name: str,
    content_type: str,
    patient_id: str,
) -> dict:
    """Upload a lab report under the lab-reports partition."""
    safe = _sanitise(original_name)
    path = f"lab-reports/{_today()}/{patient_id}_{safe}"
    url = upload_bytes(file_bytes, path, content_type, {"patientId": patient_id}, category="lab-report")
    return {"download_url": url, "storage_path": path, "file_name": original_name}


# ═══════════════════════════════════════════════════════════════════════════════
#  DOWNLOAD / SIGNED URLS
# ═══════════════════════════════════════════════════════════════════════════════

def get_download_url(storage_path: str, expiry_mins: int = SIGNED_URL_EXPIRY_MINS) -> str:
    """Generate a fresh signed download URL."""
    bucket = get_bucket()
    blob = bucket.blob(storage_path)
    if not blob.exists():
        raise FileNotFoundError(f"No file at {storage_path}")
    return blob.generate_signed_url(
        expiration=timedelta(minutes=expiry_mins),
        method="GET",
    )


# ═══════════════════════════════════════════════════════════════════════════════
#  METADATA
# ═══════════════════════════════════════════════════════════════════════════════

def get_file_metadata(storage_path: str) -> dict:
    """Retrieve metadata for a stored blob."""
    bucket = get_bucket()
    blob = bucket.blob(storage_path)
    blob.reload()
    return {
        "name": blob.name,
        "size": blob.size,
        "content_type": blob.content_type,
        "time_created": blob.time_created.isoformat() if blob.time_created else None,
        "updated": blob.updated.isoformat() if blob.updated else None,
        "custom_metadata": blob.metadata or {},
    }


def update_metadata(storage_path: str, custom_metadata: dict) -> dict:
    """Update custom metadata on an existing blob."""
    bucket = get_bucket()
    blob = bucket.blob(storage_path)
    blob.metadata = {**(blob.metadata or {}), **custom_metadata}
    blob.patch()
    return blob.metadata


# ═══════════════════════════════════════════════════════════════════════════════
#  EXISTENCE / COPY / MOVE
# ═══════════════════════════════════════════════════════════════════════════════

def file_exists(storage_path: str) -> bool:
    """Check whether a blob exists in the bucket."""
    bucket = get_bucket()
    return bucket.blob(storage_path).exists()


def copy_file(src_path: str, dest_path: str) -> str:
    """Copy a blob to a new path.  Returns the signed URL of the copy."""
    bucket = get_bucket()
    src_blob = bucket.blob(src_path)
    if not src_blob.exists():
        raise FileNotFoundError(f"Source file not found: {src_path}")
    bucket.copy_blob(src_blob, bucket, dest_path)
    dest_blob = bucket.blob(dest_path)
    return dest_blob.generate_signed_url(
        expiration=timedelta(minutes=SIGNED_URL_EXPIRY_MINS),
        method="GET",
    )


def move_file(src_path: str, dest_path: str) -> str:
    """Move (copy + delete) a blob.  Returns the signed URL of the new location."""
    url = copy_file(src_path, dest_path)
    delete_file(src_path)
    return url


# ═══════════════════════════════════════════════════════════════════════════════
#  DELETE / LIST
# ═══════════════════════════════════════════════════════════════════════════════

def delete_file(storage_path: str) -> None:
    """Delete a blob from Cloud Storage."""
    bucket = get_bucket()
    blob = bucket.blob(storage_path)
    blob.delete()


def delete_files(paths: list[str]) -> dict:
    """Delete multiple files; returns summary of successes and failures."""
    deleted, errors = [], []
    for p in paths:
        try:
            delete_file(p)
            deleted.append(p)
        except Exception as exc:
            errors.append({"path": p, "error": str(exc)})
    return {"deleted": deleted, "errors": errors}


def list_files(prefix: str) -> list[str]:
    """List all file paths under a given prefix."""
    bucket = get_bucket()
    blobs = bucket.list_blobs(prefix=prefix)
    return [blob.name for blob in blobs]


def list_files_with_metadata(prefix: str) -> list[dict]:
    """List files under a prefix with basic metadata."""
    bucket = get_bucket()
    blobs = list(bucket.list_blobs(prefix=prefix))
    results = []
    for blob in blobs:
        results.append({
            "path": blob.name,
            "size": blob.size,
            "content_type": blob.content_type,
            "time_created": blob.time_created.isoformat() if blob.time_created else None,
        })
    return results
